Add DPNP Framework Support to Selected Benchmarks

4Ran8jith opened the pull request:
integrates the DPNP framework into selected benchmarks

tbennun commented on npbench/benchmarks/contour_integral/contour_integral_dpnp.py:
```diff
@@ -0,0 +1,50 @@
+        P1 += z * X
+
+    return P0, P1
+import dpnp as np
+
+def contour_integral(NR, NM, slab_per_bc, Ham, int_pts, Y):
+    P0 = np.zeros((NR, NM), dtype=np.complex128)
+    P1 = np.zeros((NR, NM), dtype=np.complex128)
+    
+    for z in int_pts:
+        Tz = np.zeros((NR, NR), dtype=np.complex128)
+        
+        for n in range(slab_per_bc + 1):
+            zz = np.power(z, slab_per_bc / 2 - n)
+            Tz += zz * Ham[n]
+        
+        if NR == NM:
+            X = np.linalg.inv(Tz)
+        else:
+            X = np.linalg.solve(Tz, Y)
+        
+        if abs(z) < 1.0:
+            X = -X
+        
+        P0 += X
+        P1 += z * X
+
+    return P0, P1
```
the code is pasted twice

tbennun commented on npbench/benchmarks/crc16/crc16_dpnp.py:
```diff
@@ -0,0 +1,26 @@
+import numpy as np
+import dpnp
```
Dpnp is not used

tbennun commented on npbench/benchmarks/polybench/cholesky/cholesky_dpnp.py:
```diff
@@ -0,0 +1,13 @@
+#            A[i, j] /= A[j, j]
+#        A[i, i] -= np.dot(A[i, :i], A[i, :i])
+#        A[i, i] = np.sqrt(A[i, i])
+    return np.linalg.cholesky(A)
```
This is not allowed in the benchmark (this is meant to test compiler optimizations. For calling `np.linalg.cholesky`, cholesky2 exists)

tbennun commented on npbench/benchmarks/polybench/covariance/covariance_dpnp.py:
```diff
@@ -0,0 +1,10 @@
+#    cov = np.dot(np.transpose(data),data) /(float_n-1.0)
+##    for i in range(M):
+##        cov[i:M, i] = cov[i, i:M] = np.dot(data[:, i], data[:, i:M]) / (float_n - 1.0)
+    return np.cov(np.transpose(data))
```
Same comment applies here as Cholesky

tbennun commented on npbench/benchmarks/polybench/gemm/gemm_numpy.py:
```diff
@@ -4,3 +4,4 @@
 def kernel(alpha, beta, C, A, B):
 
     C[:] = alpha * A @ B + beta * C
+    return C
```
Why was a return value added to this benchmark?

tbennun commented on npbench/benchmarks/polybench/jacobi_1d/jacobi_1d_numpy.py:
```diff
@@ -6,3 +6,5 @@ def kernel(TSTEPS, A, B):
         B[1:-1] = 0.33333 * (A[:-2] + A[1:-1] + A[2:])
         A[1:-1] = 0.33333 * (B[:-2] + B[1:-1] + B[2:])
+
+    return A,B
```
Why was a return value added to this benchmark?

tbennun commented on npbench/benchmarks/polybench/k2mm/k2mm_numpy.py:
```diff
@@ -2,5 +2,5 @@
 def kernel(alpha, beta, A, B, C, D):
-
     D[:] = alpha * A @ B @ C + beta * D
+    return D
```
Why was a return value added to this benchmark?

tbennun commented on README.md:
```diff
@@ -55,6 +59,29 @@ However, you may want to install the latest version from the [GitHub repository]
+With `dpnp` it is strongly recommended to use `conda` instead of `pip` for its dependency on intel packages. 
+Refer to this 
+[LINK](https://intelpython.github.io/dpnp/quick_start_guide.html#building-for-custom-sycl-targets) to know more 
+about building custom SYCL targets or installing `dpnp` package from intel channel.
```
```suggestion
about building custom SYCL targets or installing `dpnp` package from the `intel` channel.
```

tbennun commented on README.md:
````diff
@@ -55,6 +59,29 @@ However, you may want to install the latest version from the [GitHub repository]
+
+``` bash
+$ conda env create -f environment.yml    # environment.yml contains all the right dependencies
+$ python -m pip install pygount          # Only dependency not distributed with conda
````
this should only happen after the environment is activated

tbennun commented on README.md:
````diff
@@ -55,6 +59,29 @@ However, you may want to install the latest version from the [GitHub repository]
+$ conda env create -f environment.yml    # environment.yml contains all the right dependencies
+$ python -m pip install pygount          # Only dependency not distributed with conda
+```
+If you are behind proxies, follow your system documentation. E.g. for pip you may have to specify:
+``` bash
+$ python -m pip --proxy=http://localhost:1234 install <...> # where "localhost:1234" is the value of the env var "HTTP_PROXY"
+```
````
This is unnecessary to include in the instructions. It is a pip issue rather than an NPBench issue.

tbennun reviewed (changes requested):
Thank you for your contribution! Overall it looks good. However, there are some important notes on the readme that need to be addressed, as well as other comments that I left.

4Ran8jith commented:
README.md: Updated as requested to include clarification about main.py and its purpose.

main.py: This file is designed specifically for benchmarks supporting DPNP only. It is intentionally not renamed to quickstart. A line has been added to README.md explaining this distinction.

environment.yml: Updated to reflect the latest changes.

Benchmarks: Existing benchmarks have been updated based on suggestions, and a new benchmark covariance2 has been added.

Return Statements: Modified some functions to properly use return to ensure correct copying where applicable.

Let me know if it's alright!

alexnick83 commented on main.py:
```diff
@@ -0,0 +1,72 @@
+import argparse
```
I would just remove this file from the PR for now.

alexnick83 reviewed (commented):
Looks OK to me. I suggest that you remove this main.py file; we can amend quickstart separately at a later stage.
